doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+Note that `Dataset` or rather `Variable` will be able to hold data from libraries that are *not* dependencies of this library.
+It is sufficient to forward-declare types when adding a tag for such a variable.
+*Example:
+We can store `API::Run` in `Dataset`, but the dataset library does not need to depend on `Mantid::API`.*
+
+It is currently unclear whether this library should be kept as a separate repository or be included as a new module in the Mantid repository.
+- The advantage of inclusion is the automatic use of the build server and deployment infrastructure.
+- The disadvantage is a potential coupling of versioning, see [API stability](#design-details-api-stability).
+
+If included in the Mantid repository, stand-alone packages should be provided as a light-weight alternative to downloading the full Mantid package.
+
+
+### <a name="design-details-unit-tests"></a>Unit tests
+
+The library is intended as a low-level building block and the mixed-multi-dimensional aspected lead to many corner cases.
+A comprehensive and maintainable suite of unit tests is thus essential.
+
+
+### <a name="design-details-benchmarks"></a>Benchmarks
+
+Given the intention of being a low-level building block for everything else, ensuring good performance and preventing performance regression is thus essential.
+A suite of benchmarks is to be maintained and monitored for changes.
+
+
+### <a name="design-details-sanitizers"></a>Sanitizers
+
+Static analysis and sanitization runs should be performed and verified automatically.
+Most importantly, AddressSanitizer should be used to strictly monitor for memory corruption bugs.
+
+
+### <a name="design-details-code-review"></a>Code review
+
+The same review/approval system for pull request as in Mantid should be adopted.
+However, given the scope of the project, we should require *two full reviews*, i.e., the gatekeeper should also do a thorough review, in addition to the first reviewer.
+
+
+### <a name="design-details-documentation"></a>Documentation
+
+Documentation is required for several cases:
+- Library (developer) documentation.
+- C++ API (user) documentation.
+- Python API (user) documentation.
+- Python usage examples beyond simple API documentation, e.g., how to represent complex data structures such as a `DataObjects::Workspace2D` as a `Dataset`.
+
+
+### <a name="design-details-scope-and-extension"></a>Scope and extension
+
+The design described here is seen as a nearly complete description of the scope of the dataset library.
+Extending the functionally should be avoided and should happen only after careful consideration.
+In particular, anything that is *more* specific than already supported should not be made part of this library, but instead be placed in another library that depends on this library.
+- Actually `convert` to convert units is pretty specific as it needs to connect instrument information to data.
```

Comment:
I think this could be a good idea. Perhaps the low-level stuff could be in the separate `dataset` project and then a library in mantid, that depends on `dataset` adds the higher-level operations?

Reply:
I think the unit conversion can be done without Mantid dependency, but in general I agree, there will be support libraries that depend on Mantid components. For example, we would (at least initially) reuse things like `API::Run`, and have converters from `API::MatrixWorkspace` to `Dataset`.